tizen-api/include/tizen-api-private.h

```diff
@@ -27,23 +27,111 @@
 #define __TIZEN_NNSTREAMER_API_PRIVATE_H___
 
 #include <glib.h>
+#include <gmodule.h>
 #include <gst/gst.h>
+#include "tizen-api.h"
+#include <tizen_error.h>
+#include <nnstreamer/tensor_typedef.h>
+#include <tensor_sink/tensor_sink.h>
+#include <gst/app/gstappsrc.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif /* __cplusplus */
 
+/**
+ * @brief Possible controls on elements of a pipeline.
+ */
+typedef enum {
+  NNSAPI_UNKNOWN = 0x0,
+  NNSAPI_SINK = 0x1,
+  NNSAPI_SRC = 0x2,
+  NNSAPI_VALVE = 0x3,
+  NNSAPI_SWITCH_INPUT = 0x8,
+  NNSAPI_SWITCH_OUTPUT = 0x9,
+} elementType;
+
+/**
+ * @brief Internal private representation of pipeline handle.
+ */
+typedef struct _nns_pipeline nns_pipeline;
+
+/**
+ * @brief An element that may be controlled individually in a pipeline.
+ */
+typedef struct _element {
+  GstElement *element; /**< The Sink/Src/Valve/Switch element */
+  nns_pipeline *pipe; /**< The main pipeline */
+  const char *name;
+  elementType type;
+  GstPad *src;
+  GstPad *sink; /**< Unref this at destroy */
+  GstTensorsInfo tensorsinfo;
+  size_t size;
+
+  GList *handles;
+  GMutex lock; /**< Lock for internal values */
+} element;
+
 /**
  * @brief Internal private representation of pipeline handle.
  * @detail This should not be exposed to applications
  */
-typedef struct _nns_pipeline {
+struct _nns_pipeline {
+  GstElement *element;    /**< The pipeline itself (GstPipeline) */
+  GMutex lock;            /**< Lock for pipeline operations */
+  GHashTable *namednodes; /**< hash table of "element"s. */
+};
+
+/**
+ * @brief Internal private representation of sink handle of GstTensorSink
+ * @detail This represents a single instance of callback registration. This should not be exposed to applications.
+ */
+typedef struct _nns_sink {
+  nns_pipeline *pipe; /**< The pipeline, which is the owner of this nns_sink */
+  GstTensorSink *element;
```

Comment:
Do we need to use the struct 'GstTensorSink' 'GstAppSrc' ?

Reply:
No, we need element, not GstObject. I'll backport it from tizen-api/0.0.6 branch.